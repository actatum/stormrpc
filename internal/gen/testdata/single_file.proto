syntax = "proto3";

package test;

option go_package = "./test;prototest";

service Greeter {
    rpc SayHello (HelloRequest) returns (HelloReply) {}
}

message HelloRequest {
    string name = 1;
}

message HelloReply {
    string message = 1;
}