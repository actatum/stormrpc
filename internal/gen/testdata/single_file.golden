// Code generated by protoc-gen-stormrpc. DO NOT EDIT.

package prototest

import (
	context "context"
	fmt "fmt"
	stormrpc "github.com/actatum/stormrpc"
)

// GreeterClient is the client API for Greeter service.
type GreeterClient interface {
	SayHello(ctx context.Context, in *HelloRequest, opts ...stormrpc.CallOption) (*HelloReply, error)
}

type greeterClient struct {
	c *stormrpc.Client
}

func NewGreeterClient(c *stormrpc.Client) GreeterClient {
	return &greeterClient{c}
}

func (c *greeterClient) SayHello(ctx context.Context, in *HelloRequest, opts ...stormrpc.CallOption) (*HelloReply, error) {
	var out HelloReply
	r, err := stormrpc.NewRequest("rpc.Greeter.SayHello", in, stormrpc.WithEncodeProto())
	if err != nil {
		return nil, err
	}

	resp := c.c.Do(ctx, r, opts...)
	if resp.Err != nil {
		return nil, resp.Err
	}

	if err = resp.Decode(&out); err != nil {
		return nil, err
	}

	return &out, nil
}

// GreeterServer is the server API for Greeter service.
type GreeterServer interface {
	SayHello(context.Context, *HelloRequest) (*HelloReply, error)
}

func RegisterGreeterServer(s *stormrpc.Server, srv GreeterServer) {
	for _, handler := range greeterHandlers {
		handler.SetService(srv)
		s.Handle(handler.Route(), handler.HandlerFunc())
	}
}

type _Greeter_SayHello_Handler struct {
	route string
	svc   interface{}
}

func (h *_Greeter_SayHello_Handler) HandlerFunc() stormrpc.HandlerFunc {
	return func(ctx context.Context, r stormrpc.Request) stormrpc.Response {
		var in HelloRequest
		if err := r.Decode(&in); err != nil {
			return stormrpc.NewErrorResponse(r.Reply, fmt.Errorf("error decoding request"))
		}

		out, err := h.svc.(GreeterServer).SayHello(ctx, &in)
		if err != nil {
			return stormrpc.NewErrorResponse(r.Reply, err)
		}

		resp, err := stormrpc.NewResponse(r.Reply, out, stormrpc.WithEncodeProto())
		if err != nil {
			return stormrpc.NewErrorResponse(r.Reply, err)
		}

		return resp
	}
}

func (h *_Greeter_SayHello_Handler) Route() string {
	return h.route
}

func (h *_Greeter_SayHello_Handler) SetService(svc interface{}) {
	h.svc = svc
}

var greeterHandlers = []handler{
	&_Greeter_SayHello_Handler{route: "rpc.Greeter.SayHello"},
}

type handler interface {
	Route() string
	HandlerFunc() stormrpc.HandlerFunc
	SetService(interface{})
}
